src/armature/CostFunction.ts

```diff
@@ -0,0 +1,184 @@
+import { coord } from '../calder';
+import { AABB } from '../geometry/BakedGeometry';
+import { vec3From4 } from '../math/utils';
+import { Mapper } from '../utils/mapper';
+import { CostFn, GeneratorInstance } from './Generator';
+import { Model } from './Model';
+import { GeometryNode, Node } from './Node';
+
+import { vec3, vec4 } from 'gl-matrix';
+import { range } from 'lodash';
+
+export type ForcePoint = {
+    point: coord;
+    influence: number;
+};
+
+type Grid = { [key: string]: true };
+
+export namespace CostFunction {
+    /**
+     * Use points with positive and negative influence to control generation.
+     *
+     * @param {ForcePoint[]} points The points of influence, where negative influence means the
+     * point reduces the overall cost when nodes get close.
+     * @returns {CostFn} The resulting cost function.
+     */
+    export function forces(points: ForcePoint[]): CostFn {
+        const vectors = points.map((forcePoint: ForcePoint) => {
+            return {
+                vector: Mapper.coordToVector(forcePoint.point),
+                influence: forcePoint.influence
+            };
+        });
+
+        return (instance: GeneratorInstance, added: Node[]) => {
+            // Out of the added nodes, just get the geometry nodes
+            const addedGeometry: GeometryNode[] = [];
+            added.forEach((n: Node) =>
+                n.geometryCallback((node: GeometryNode) => {
+                    addedGeometry.push(node);
+                })
+            );
+
+            let totalCost = instance.getCost();
+
+            // For each added shape and each influence point, add the resulting cost to the
+            // instance's existing cost.
+            addedGeometry.forEach((node: GeometryNode) => {
+                const localToGlobalTransform = node.localToGlobalTransform();
+                const globalPosition = vec3From4(
+                    vec4.transformMat4(
+                        vec4.create(),
+                        vec4.fromValues(0, 0, 0, 1),
+                        localToGlobalTransform
+                    )
+                );
+
+                vectors.forEach((point: { vector: vec3; influence: number }) => {
+                    // Add cost relative to the point's influence, and inversely proportional
+                    // to the distance to the point
+                    totalCost +=
+                        point.influence /
+                        vec3.length(vec3.sub(vec3.create(), point.vector, globalPosition));
```

Comment:
I think that's worth trying, although I think changes to the cost functions themselves I'm going to do in another PR so that the boilerplate can be merged so it's easier for everyone else to test things